Add property DoesBodyEndWithCompoundStmt

PappasBrent opened the pull request:
Adds the syntactic property `DoesBodyEndWithCompoundStmt` to keep track of macros which expand to non-expression statements that end in a compound statement.

Downstream translation tools should not translate any macro that satisfy this property to a function.
The reason is that a macro that expands to a non-expression statement ending in a compound statement need not be followed by a semicolon at its callsites; however, if it were instead defined as a function, then its callsites would each need to be followed by a semicolon.
The only way to correctly translate such a macro would be to insert a semicolon after each of its callsites in addition to translating the macro's definition into a function definition.

Closes #73

PappasBrent committed:
Add endsWithCompound() function

Add endsWithCompound() function to check if a given statement ends with
a compound statement.

PappasBrent committed:
Add property DoesBodyEndWithCompoundStmt

Add the syntactic property DoesBodyEndWithCompoundStmt to check if a
macro expands to a statement that ends with a compound statement. Macros
that satisfy this property do not need to be followed by a semicolon at
their callsites, so translation tools like MerC may wish to avoid
translating them to functions since function calls must be followed by
semicolons at their callsites (assuming they are not part of a larger
expression, which macros that end with a compound statement will not
be).

PappasBrent committed:
Add test for DoesBodyEndWithCompoundStmt

SilverMight commented on lib/ASTUtils.cc:
```diff
@@ -249,4 +250,24 @@ std::set<const clang::Stmt *> subtrees(const clang::Stmt *ST) {
+            return endsWithCompound(Else);
+        } else {
+            return endsWithCompound(If->getThen());
+        }
```
I think this may need to handle a switch statement, I'm able to create a case like this:
```c
#define TEST(X) switch(X) {case 1: ;}

int main(int argc, char *argv[])
{

        TEST(1)
        return 0;
}
~```
`DoesBodyEndWithCompoundStmt` will be false in this case, and MerC  translates the macro and causes a compilation error.

PappasBrent replied:
Ah nice catch, I just added a fix

PappasBrent committed:
Add switch stmts to endsWithCompound()

PappasBrent committed:
Add switch statements to test case

PappasBrent committed:
Only find clang-17

Only find clang-17 since that is the specific Clang version we support.

PappasBrent commented:
I added a slight change to the top-level CMakeLists.txt to only search for the clang-17 compiler specifically.
Before, we were searching for the default clang compiler and then falling back to clang-17 in case a default clang compiler was not found.
I just updated to clang-19 though, and I realized that this is probably not the behavior that we want, because when my CMake defaulted to clang-19 as the compiler, Maki compiled but would emit nothing, with no warning messages either.
So I think we should explicitly find clang-17 to avoid subtle errors like this in the future.

SilverMight commented:
> I added a slight change to the top-level CMakeLists.txt to only search for the clang-17 compiler specifically. Before, we were searching for the default clang compiler and then falling back to clang-17 in case a default clang compiler was not found. I just updated to clang-19 though, and I realized that this is probably not the behavior that we want, because when my CMake defaulted to clang-19 as the compiler, Maki compiled but would emit nothing, with no warning messages either. So I think we should explicitly find clang-17 to avoid subtle errors like this in the future.

Hm, I meant to target the clang 17 API with the `find_package(Clang ${MAKI_CLANG_VERSION} REQUIRED CONFIG)` line, when I try to run it on my machine (with only clang 18) it complains about not finding the correct version:

```
CMake Error at CMakeLists.txt:19 (find_package):
  Could not find a configuration file for package "Clang" that is compatible
  with requested version "17".

  The following configuration files were considered but not accepted:

    /usr/lib/cmake/clang/ClangConfig.cmake, version: 18.1.8
    /lib/cmake/clang/ClangConfig.cmake, version: 18.1.8
```

I guess theoretically, you could have multiple versions on one system and have the Maki script use the wrong compiler version. So I guess it's a good idea to be more specific.
